experimental/lib/Support/FtdSupport.cpp

```diff
@@ -14,24 +14,91 @@
 
 #include "experimental/Support/FtdSupport.h"
 #include "mlir/Dialect/ControlFlow/IR/ControlFlowOps.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
 #include <unordered_set>
 
 using namespace mlir;
 using namespace dynamatic;
-using namespace experimental::boolean;
+using namespace dynamatic::experimental;
 
-/// Recursive function which allows to obtain all the paths from block "start"
-/// to block "end" using a DFS, possibly traversing "blockToTraverse" and not
-/// traversing "blocksToAvoid".
+std::string ftd::BlockIndexing::getBlockCondition(Block *block) const {
+  return "c" + std::to_string(getIndexFromBlock(block));
+}
+
+ftd::BlockIndexing::BlockIndexing(Region &region) {
+  mlir::DominanceInfo domInfo;
+
+  // Create a vector with all the blocks
+  SmallVector<Block *> allBlocks;
+  for (Block &bb : region.getBlocks())
+    allBlocks.push_back(&bb);
+
+  // Sort the vector according to the dominance information, so that a block
+  // comes before each dominators.
+  std::sort(allBlocks.begin(), allBlocks.end(),
+            [&](Block *a, Block *b) { return domInfo.dominates(a, b); });
+
+  // Associate a smalled index in the map to the blocks at higer levels of the
+  // dominance tree
+  for (auto [blockID, bb] : llvm::enumerate(allBlocks)) {
+    indexToBlock.insert({blockID, bb});
+    blockToIndex.insert({bb, blockID});
+  }
+}
+
+Block *ftd::BlockIndexing::getBlockFromIndex(unsigned index) const {
+  auto it = indexToBlock.find(index);
+  return (it == indexToBlock.end()) ? nullptr : it->getSecond();
```

Comment:
just `return indexToBlock.lookup(index)`, which you can directly do in the header.